src/handle_list.rs

```diff
@@ -0,0 +1,227 @@
+    pub fn new_entry(&self) -> Arc<AtomicUsize> {
+        let count = Arc::new(AtomicUsize::new(0));
+
+        self.add_counter(count.clone());
+        count
+    }
+    fn add_counter(&self, count: Arc<AtomicUsize>) {
+        let n_node = Box::new(ListEntry {
+            data: count,
+            next: core::ptr::null(),
+        });
+        let n_node_ptr = Box::into_raw(n_node);
+
+        let mut current_head = self.inner.head.load(Ordering::SeqCst);
+        loop {
+            // Safety
+            // This is save, because we have not stored the Ptr elsewhere so we have exclusive
+            // access.
+            // The Ptr is also still valid, because we never free Entries on the List
+            unsafe { (*n_node_ptr).next = current_head };
+
+            // Attempt to add the Entry to the List by setting it as the new Head
+            match self.inner.head.compare_exchange(
+                current_head,
+                n_node_ptr,
+                Ordering::SeqCst,
+                Ordering::SeqCst,
+            ) {
+                Ok(_) => return,
+                Err(n_head) => {
+                    // Store the found Head-Ptr to avoid an extra load at the start of every loop
+                    current_head = n_head;
+                }
+            }
+        }
+    }
+
+    /// Creates a new Snapshot of the List at this Point in Time
+    pub fn snapshot(&self) -> ListSnapshot {
+        ListSnapshot {
+            head: self.inner.head.load(Ordering::SeqCst),
+            _list: self.inner.clone(),
+        }
+    }
+
+    /// Inserts the Items of the Iterator, but in reverse order
+    #[cfg(test)]
+    pub fn extend<I>(&self, iter: I)
+    where
+        I: IntoIterator<Item = Arc<AtomicUsize>>,
+    {
+        for item in iter.into_iter() {
+            self.add_counter(item);
+        }
+    }
+}
+
+impl Default for HandleList {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+impl Debug for HandleList {
+    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
+        // TODO
+        // Figure out how exactly we want the Debug output to look
+        write!(f, "HandleList")
+    }
+}
+impl Clone for HandleList {
+    fn clone(&self) -> Self {
+        Self {
+            inner: Arc::clone(&self.inner),
+        }
+    }
+}
+
+impl ListSnapshot {
+    /// Obtain an iterator over the Entries in this Snapshot
+    pub fn iter(&self) -> SnapshotIter {
+        SnapshotIter { current: self.head }
+    }
+}
+
+impl Iterator for SnapshotIter {
+    // TODO
+    // Maybe don't return an owned Value here
+    type Item = Arc<AtomicUsize>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.current.is_null() {
+            return None;
+        }
+
+        // Safety
+        // The Ptr is not null, because of the previous if-statement.
+        // The Data is also not freed, because we never free Entries on the List.
```

Comment:
This is no longer true.

Reply:
Updated it with the correct reasoning now, though not quite sure on the wording I used